File: lib/ts-sdk/examples/custom-filters.ts
/**
 * Scenario 5 - custom fields + a registered custom filter.
 *
 * AUTHOR: register a custom filter on the widgets search route via `routes`.
 * CONSUMER: get a typed client and call `widgets.search({ filters })`: the standard `color`
 * key routes to the top level, the registered `region` and an ad hoc `tier` pass through to
 * `customFilters`, and rows are typed as the plugin's Widget model.
 *
 * Stubs `globalThis.fetch` so it runs offline.
 * Run: `pnpm dlx tsx examples/custom-filters.ts` (or `pnpm run examples` for all scenarios).
 */

import { definePlugin, f } from "../src";
import { check, widgetCustomFields } from "./source";

// --- Author -----------------------------------------------------------------------------
export const routesPlugin = definePlugin({
  meta: { name: "widget routes plugin", version: "0.1.0", sourceSystem: "acme-widgets" },
  schemas: { Widget: { customFields: widgetCustomFields } },
  routes: {
    widgets: {
      // `color` is a protocol DEFAULT filter (not declared here); `region` is the custom
      // filter this plugin registers. Both registered and ad hoc keys nest under
      // `customFilters` in the request body.
      search: { filters: { region: { filterType: "stringArray" } } },
    },
  },
});

// --- Consumer ---------------------------------------------------------------------------
let lastSearchBody: unknown;

function stubFetch(): void {
  globalThis.fetch = (async (input: string | URL | Request, init?: RequestInit) => {
    const url = typeof input === "string" ? input : input instanceof URL ? input.href : input.url;
    if (url.endsWith("/widgets/search")) {
      lastSearchBody = init?.body ? JSON.parse(String(init.body)) : undefined;
    }
    return new Response(
      JSON.stringify({
        status: 200,
        message: "ok",
        items: [
          { id: "123e4567-e89b-42d3-a456-426614174001", name: "Red", color: "red", weight: 5 },
        ],
        paginationInfo: { page: 1, pageSize: 1, totalItems: 1, totalPages: 1 },
        sortInfo: { sortBy: "id", sortOrder: "asc" },
        filterInfo: { filters: {} },
      }),
      { status: 200, headers: { "Content-Type": "application/json" } }
    );
  }) as typeof fetch;
}

export async function demo(): Promise<void> {
  console.log("Scenario 5 - custom fields + a registered custom filter");
  stubFetch();

  const client = routesPlugin.getClient({ baseUrl: "https://api.example.org" });
  const result = await client.widgets.search({
    page: 1,
    filters: {
      color: f.eq("red"), // standard -> top level
      region: f.in(["PA", "NJ"]), // registered custom -> customFilters
      tier: f.eq("gold"), // ad hoc -> customFilters (passthrough)
    },
  });
  // Rows are typed as the plugin's Widget model, no call-site annotations.
  const firstName: string | undefined = result.items[0]?.ok ? result.items[0].data.name : undefined;
  check("search returned a typed row", firstName === "Red");

  const filters = (lastSearchBody as { filters?: Record<string, unknown> })?.filters ?? {};
  const customFilters = (filters.customFilters as Record<string, unknown>) ?? {};
  check("standard color routed to the top level", "color" in filters && !("region" in filters));
  check(
    "region (registered) + tier (ad hoc) -> customFilters",
    "region" in customFilters && "tier" in customFilters
  );
}

demo().catch((err) => {
  console.error(err);
  process.exitCode = 1;
});

File: lib/py-sdk/examples/custom_filters.py
"""Scenario 5 - custom fields + a registered custom filter.

AUTHOR: extend the resource's standard-filters TypedDict with a custom filter and register it
on the route via ``Routes`` / ``ResourceRoutes`` / ``RouteFilters``.
CONSUMER: get a typed client and call ``widgets.search(filters=...)``: the standard ``color``
key routes to the top level, the registered ``region`` and an ad hoc ``tier`` pass through to
``customFilters``, and rows are typed as the plugin's Widget model.

Run: ``python -m examples.custom_filters`` (or ``python -m examples`` for all scenarios).
"""

from __future__ import annotations

import json
from typing import Any, assert_type

import httpx

from common_benefits_sdk.client import Config, SearchResult
from common_benefits_sdk.extensions import (
    PluginMeta,
    PluginSchemas,
    ResourceRoutes,
    RouteFilters,
    Routes,
    define_plugin,
    schema,
)
from common_benefits_sdk.schemas.filters import StringArray, WidgetFilters, f
from common_benefits_sdk.schemas.models import WidgetCommon

from .source import WidgetFields


# --- Author -----------------------------------------------------------------------------
# Extend the standard WidgetFilters with the custom filter, then name it in routes so
# `region` autocompletes (and is value-typed) on widgets.search.
class WidgetSearchFilters(WidgetFilters, total=False):
    region: StringArray


routes_plugin = define_plugin(
    schemas=PluginSchemas(Widget=schema(common_schema=WidgetCommon[WidgetFields])),
    routes=Routes(widget=ResourceRoutes(search=RouteFilters[WidgetSearchFilters]())),
    meta=PluginMeta(name="widget routes plugin", source_system="acme-widgets"),
)


# --- Consumer ---------------------------------------------------------------------------
def _page() -> dict:
    return {
        "items": [{"id": "w-1", "name": "Red", "color": "red", "weight": 5.0}],
        "paginationInfo": {"page": 1, "pageSize": 1, "totalItems": 1, "totalPages": 1},
    }


def demo() -> None:
    print("Scenario 5 - custom fields + a registered custom filter")

    captured: dict[str, Any] = {}

    def handler(request: httpx.Request) -> httpx.Response:
        captured["filters"] = json.loads(request.content)["filters"]
        return httpx.Response(200, json=_page())

    config = Config(
        base_url="https://example.test", transport=httpx.MockTransport(handler)
    )
    with routes_plugin.get_client(config) as client:
        result = client.widgets.search(
            filters={
                "color": f.eq("red"),  # standard -> top level
                "region": f.in_(["PA", "CA"]),  # registered custom -> customFilters
                "tier": f.eq("gold"),  # ad hoc -> customFilters (passthrough)
            }
        )
        # Rows are typed as the plugin's Widget model, no call-site annotations.
        assert_type(result, SearchResult[WidgetCommon[WidgetFields]])

    filters = captured["filters"]
    top_level = "color" in filters and "region" not in filters
    passed_through = set(filters.get("customFilters", {})) == {"region", "tier"}
    print(
        f"  [{'PASS' if top_level else 'FAIL'}] standard color routed to the top level"
    )
    print(
        f"  [{'PASS' if passed_through else 'FAIL'}] region (registered) + tier (ad hoc) -> customFilters"
    )


if __name__ == "__main__":
    demo()
